filecache/file_cache_path.py

```diff
@@ -1122,146 +1191,585 @@
+                        continue
+                    if dir_only:
+                        yield from select_next(entry, exists=True)
+                    else:
+                        yield entry
+        return select_wildcard
+
+    def recursive_selector(self,
+                           part: str,
+                           parts: list[str]) -> Callable[[str, bool], Generator[FCPath]]:
+        """Returns a function that selects a given path and all its children,
+        recursively, filtering by pattern.
+        """
+        # Optimization: consume following '**' parts, which have no effect.
+        while parts and parts[-1] == '**':
+            parts.pop()
+
+        # Optimization: consume and join any following non-special parts here,
+        # rather than leaving them for the next selector. They're used to
+        # build a regular expression, which we use to filter the results of
+        # the recursive walk. As a result, non-special pattern segments
+        # following a '**' wildcard don't require additional filesystem access
+        # to expand.
+        match = None if part == '**' else self.compile(part)
+        dir_only = bool(parts)
+        select_next = self.selector(parts)
+
+        def select_recursive(path: str,
+                             exists: bool = False) -> Generator[FCPath]:
+            if path and path[-1] != '/':
+                path = f'{path}/'
+            match_pos = len(str(path))
+            if match is None or match(str(path), match_pos):
+                yield from select_next(path, exists)
+            stack = [path]
+            while stack:
+                yield from select_recursive_step(stack, match_pos)
+
+        def select_recursive_step(stack: list[str],
+                                  match_pos: int) -> Generator[Any]:
+            path = stack.pop()
+            entries = list(FCPath(path, copy_from=self.copy_from).iterdir_type())
+            for entry, is_dir in entries:
+                if is_dir or not dir_only:
+                    if match is None or match(str(entry), match_pos):
+                        if dir_only:
+                            yield from select_next(entry, exists=True)
+                        else:
+                            # Optimization: directly yield the path if this is
+                            # last pattern part.
+                            yield entry
+                    if is_dir:
+                        stack.append(entry.path)
+
+        return select_recursive
+
+    def select_exists(self,
+                      path: str,
+                      exists: bool = False) -> Generator[str]:
+        """Yields the given path, if it exists.
+        """
+        if exists:
+            # Optimization: this path is already known to exist, e.g. because
+            # it was returned from os.iterdir(), so we skip calling exists().
+            yield path
+        else:
+            if FCPath(path, copy_from=self.copy_from).exists():
```

Comment:
<div>


<div id="suggestion">
<div id="issue"><b>Consider validating copy_from parameter first</b></div>
<div id="fix">

Consider validating the `copy_from` parameter before passing it to `FCPath` constructor. The `copy_from` parameter could potentially be `None` or invalid.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<blockquote>Check the AI-generated fix before applying</blockquote>
<div id="code">


```suggestion
            if self.copy_from is not None and FCPath(path, copy_from=self.copy_from).exists():
```

</div>
</details>
</div>


<small><i>Code Review Run <a href=https://github.com/SETI/rms-filecache/pull/32#issuecomment-2558333795>#dae64b</a></i></small>
</div>

---
Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged